hw/dv/dpi/gpiodpi/gpiodpi.c

```diff
@@ -21,83 +21,238 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+// This file does a lot of bit setting and getting; these macros are intended to
+// make that a little more readable.
+#define GET_BIT(word, bit_idx) ((word) >> bit_idx) & 1
+#define SET_BIT(word, bit_idx) word |= (1 << bit_idx);
+#define CLR_BIT(word, bit_idx) word &= ~(1 << bit_idx);
+
+typedef struct gpiodpi_ctx {
+  // The number of pins we're driving.
+  int n_bits;
+
+  // The last known value of the pins, in little-endian order.
+  uint32_t driven_pin_values;
+
+  // File descriptors and paths for the device-to-host and host-to-device
+  // FIFOs.
+  int dev_to_host_fifo;
+  char dev_to_host_path[PATH_MAX];
+  int host_to_dev_fifo;
+  char host_to_dev_path[PATH_MAX];
+} gpiodpi_ctx_t;
+
+/**
+ * Creates a new UNIX FIFO file at |path_buf|, and opens it with |flags|.
+ *
+ * @return a file descriptor for the fifo, or -1 if any syscall failed.
+ */
+static int open_fifo(char *path_buf, int flags) {
+  int fifo_status = mkfifo(path_buf, 0644);
+  if (fifo_status != 0) {
+    fprintf(stderr, "GPIO: Unable to create FIFO at %s: %s\n", path_buf,
+            strerror(errno));
+    return -1;
+  }
+
+  int fd = open(path_buf, flags);
+  if (fd < 0) {
+    // Delete the fifo we created; ignore errors.
+    unlink(path_buf);
+    fprintf(stderr, "GPIO: Unable to open FIFO at %s: %s\n", path_buf,
+            strerror(errno));
+    return -1;
+  }
+
+  return fd;
+}
+
+/**
+ * Print out a useage message for the GPIO interface.
+ *
+ * @arg rfifo the path to the "read" side (w.r.t the host).
+ * @arg wfifo the path to the "write" side (w.r.t the host).
+ * @arg n_bits the number of pins supported.
+ */
+static void print_usage(char *rfifo, char *wfifo, int n_bits) {
+  printf("\n");
+  printf("GPIO: FIFO pipes created at %s and %s for %d-bit wide GPIO.\n", rfifo,
+         wfifo, n_bits);
+  printf("GPIO: To read the values of the pins relative to ground, run\n");
+  printf("$ cat %s  # '0' low, '1' high, 'X' floating\n", rfifo);
+  printf("GPIO: To drive the pins, run a command like\n");
+  printf("$ echo 'h0a l1f' > %s  # Pull the 10th pin high, and the 31st low.\n",
```

Comment:
- the 10th pin is pin 9 when starting to count from 0. 
- Pins are typically numbered in decimal, not hex. (People talk about pin 10, not pin a.) It would increase usability if the tool accepts pin indices in decimal.